README.rst
===============================
pyglobalgoals
===============================

.. .. image:: https://img.shields.io/travis/westurner/pyglobalgoals.svg
..         :target: https://travis-ci.org/westurner/pyglobalgoals
.. ..
.. .. image:: https://img.shields.io/pypi/v/pyglobalgoals.svg
..        :target: https://pypi.python.org/pypi/pyglobalgoals


pyglobalgoals is a Python package, Python module, and a set of Python Jupyter notebooks for working with JSON-LD, RDFa, schema.org and The Global Goals For Sustainable Development (#GlobalGoals #GGs #SDG #SDGs) #GG17

* Free software: BSD license

.. * Documentation: https://pyglobalgoals.readthedocs.org.
.. Features
.. --------
.. 
.. * TODO


Global Goals
==============
| Homepage: http://www.globalgoals.org/
| Twitter: https://twitter.com/TheGlobalGoals
| Instagram: https://instagram.com/TheGlobalGoals/
| Facebook: https://www.facebook.com/globalgoals.org
| YouTube: https://www.youtube.com/channel/UCRfuAYy7MesZmgOi1Ezy0ng/
| Hashtag: **#GlobalGoals**
| - https://twitter.com/hashtag/GlobalGoals
| - https://instagram.com/explore/tags/GlobalGoals/
| - https://www.facebook.com/hashtag/GlobalGoals
| Hashtag: #TheGlobalGoals
| - https://twitter.com/hashtag/TheGlobalGoals
| - https://instagram.com/explore/tags/TheGlobalGoals/
| - https://www.facebook.com/hashtag/TheGlobalGoals


pyglobalgoals
===============

| Homepage: https://github.com/westurner/pyglobalgoals
| Src: https://github.com/westurner/pyglobalgoals
| Download: https://github.com/westurner/pyglobalgoals/releases

* `<https://github.com/westurner/pyglobalgoals/blob/master/notebooks/globalgoals-pyglobalgoals.py.ipynb>`__
* `<https://github.com/westurner/pyglobalgoals/blob/develop/notebooks/globalgoals-pyglobalgoals.py.ipynb>`__


License
===========
* pyglobalgoals: BSD License
  ``LICENSE``
* GlobalGoals explanations and descriptions:
  Creative Commons Attribution-ShareAlike 3.0
  https://creativecommons.org/licenses/by-sa/3.0/
  ``LICENSE.cc-by-sa-3.0.html``
* GlobalGoals images (provided by Getty Images):
  http://www.globalgoals.org/asset-licence/
  ``LICENSE.globalgoals-asset-license.html``


HISTORY.rst
.. :changelog:

History
-------



release/0.2.5 (2016-03-08 17:18:54 -0600)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.2.4..release/0.2.5

* Merge tag 'v0.2.4' into develop [ff318c7]
* DOC: README.rst: [schema:]description [cba7b86]
* BUG: glogbalgoals-pyglobalgoals.py.{py,ipynb}: RDFa prefix= attribute newlines [561c574]
* Merge remote-tracking branch 'origin/develop' into develop [64a7c50]
* BLD: Makefile: jupyter-notebook nbconvert [8ec96bf]
* DOC,BUG: globalgoals-pyglobalgoals.py.ipynb, notebooks/data: title, links, get_goal_tile_data, new img URLs [d99cc93]
* RLS,DOC: globalgoals-pyglobalgoals.py.ipynb: #GG17, v0.2.5, restart/clear/runall [3d3722f]
* RLS: setup.py, __init__.py: v0.2.5 [6b7c8aa]


v0.2.4 (2016-01-02 13:06:54 -0600)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.2.3..v0.2.4

* Merge tag 'v0.2.3' into develop [c4e3113]
* REF: pyglobalgoals-svg.ipynb: from globalgoals-pyglobalgoals.ipynb [92b9afd]
* RLS: setup.py, __init__.py: v0.2.4 [fe9af0e]
* DOC: HISTORY.rst: output from ``git-changelog.py -r release/0.2.4`` [ee7bd20]
* Merge branch 'release/0.2.4' [6094bbd]


v0.2.3 (2015-12-04 07:55:38 -0600)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.2.2..v0.2.3

* Merge tag 'v0.2.2' into develop [4d5ab5f]
* BUG,BLD: setup.py: jinja2, requests [8266bd7]
* BUG,ENH: .ipynb, globalgoals.jsonld: description.@type=schema:Text, pip install get_image_size [f7fcc57]
* Merge branch 'develop' of ssh://github.com/westurner/pyglobalgoals into develop [c1f3935]
* CLN: .gitignore, notebooks/.ipynb_checkpoints: add to .gitignore [a85c726]
* RLS: setup.py, __init__.py: v0.2.3 [2704fd7]
* Merge branch 'release/0.2.3' [15aa97e]


v0.2.2 (2015-10-19 22:35:41 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.2.1..v0.2.2

* Merge tag 'v0.2.1' into develop [abd1c88]
* ENH: .ipynb: Generate globalgoals.[html,jsonld,md,rdfa.html5,rst] [996ef6b]
* BLD: notebooks/globalgoals-pyglobalgoals.py.py: Manually update 'Download as' > 'Python (.py)' output [d9675d8]
* RLS: setup.py, __init__.py: version = '0.2.2' [c670962]
* Merge branch 'release/0.2.2' [6119e83]


v0.2.1 (2015-09-22 03:02:44 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.2.0..v0.2.1

* Merge tag 'v0.2.0' into develop [12f60a7]
* BLD: Makefile, environment.yml, requirements.txt, setup.py: dependencies [ba91d15]
* ENH: env..yml, requirements.txt: add version_information extension [b10da91]
* ENH,CLN: globalgoals-pyglobalgoals.py.ipynb: version_information, dependencies, img download output seq [4333c82]
* REF: version_information, link to v0.2.1 [a71b846]
* DOC: README.rst: links from notebooks/globalgoals-pyglobalgoals.py.ipynb [74eba5f]
* DOC: README.rst: links to notebooks/globalgoals-pyglobalgoals.py.ipynb [99e2abb]
* Merge branch 'release/0.2.1' [6a4fee7]


v0.2.0 (2015-09-22 01:50:27 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.1.2..v0.2.0

* Merge tag 'v0.1.2' into develop [118d3f2]
* REF,DOC: links, functions, RST, Markdown, RDFa [e357f3c]
* RLS: setup.py, __init__.py: v0.2.0 [53bd0f9]
* Merge branch 'release/0.2.0' [dc0acf2]


v0.1.2 (2015-09-21 20:03:36 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.1.1..v0.1.2

* Merge tag 'v0.1.1' into develop [8eec0dc]
* REF: notebooks/globalgoals-tweepy.py.ipynb: globalgoals-pyglobalgoals.py.ipynb [52a5732]
* REF: globalgoals-tweepy.py -> globalgoals-pyglobalgoals.py, globalgoals.jsonld indent=2 [26a993e]
* RLS: setup.py, __init__.py: v0.1.2 [a1550f9]
* Merge branch 'release/0.1.2' [2fbba76]


v0.1.1 (2015-09-21 19:29:49 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' v0.1.0..v0.1.1

* Merge tag 'v0.1.0' into develop [4a63ed1]
* DOC: globalgoals-tweepy.py.ipynb: link, {restart, run all} [d31218a]
* RLS: setup.py, ./__init__.py: v0.1.1 [52a653b]
* Merge branch 'release/0.1.1' [0d1c99f]


v0.1.0 (2015-09-21 19:13:55 -0500)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
::

   git log --reverse --pretty=format:'* %s [%h]' ac91c7d..v0.1.0

* BLD: setup.py, requirements.txt: pip install tweepy [5601d7a]
* BLD: setup.py, requirements.txt: beautifulsoup4 [53697b9]
* DOC: notebooks/globalgoals-tweepy.py.ipynb: get goals from http://www.globalgoals.org [8073512]
* DOC: notebooks/globalgoals-tweepy.py.py: get goals from http://www.globalgoals.org [b95d72e]
* DOC: nb/globalgoals-tweepy.py.ipynb: examples [8b57bbe]
* DOC: globalgoals-tweepy.py.ipyng: markdown headings, link to src; clear, restart, run all [ab2bf11]
* DOC: README.rst: http://www.globalgoals.org/ [411c6b7]
* Merge branch 'feature/tweepy_demo' into develop [e37b18b]
* DOC: README.rst [de60156]
* ENH: notebooks/data/images, IPython.display. [f0a0d2a]
* Merge branch 'release/0.1.0' [8285337]



tests/test_pyglobalgoals.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_pyglobalgoals
----------------------------------

Tests for `pyglobalgoals` module.
"""

import unittest

from pyglobalgoals import pyglobalgoals


class TestPyglobalgoals(unittest.TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


pyglobalgoals-svg.py

# coding: utf-8

# # pyglobalgoals-svg

# * | Src:
# * | Nbviewer:

# In[3]:

get_ipython().system(u'pip install svgwrite')
# https://svgwrite.readthedocs.org/en/latest/overview.html
import svgwrite

get_ipython().system(u'pip install -e git+https://github.com/scardine/image_size@324affe#egg=image-size')
import get_image_size


# In[ ]:

dwg = svgwrite.Drawing(height=800, width=600)
print(('dwg', dwg))
img = svgwrite.image.Image(href='./data/images/partnership.png')
print(('img', img))
dwg.add(img)
print(('dwg', dwg))
print(('elements', dwg.elements))
for e in dwg.elements:
    print((e, e.tostring()))
print(('dwg_svg', dwg.tostring()))

def prettify_xml(obj):
    return (bs4.BeautifulSoup(dwg.tostring(), features='xml').prettify())
print("")
print(prettify_xml(dwg.tostring()))


# In[4]:

import glob
files = glob.glob('./data/images/*.png')
for img_path in files:
    md = get_image_size.get_image_metadata(img_path)
    print((img_path, md.height, md.width)) #, md ))


# In[ ]:





notebooks/globalgoals-tweepy.py.py

# coding: utf-8

# In[1]:

import tweepy
import bs4
import requests

import pyglobalgoals.pyglobalgoals as gg


# In[2]:

url = "http://www.globalgoals.org/"
req = requests.get(url)
print(req)
print(sorted(dir(req)))


# In[3]:

if not req.ok:
    raise Exception(req)
content = req.content
print(content[:20])


# In[4]:

bs = bs4.BeautifulSoup(req.content)
print(bs.prettify())


# In[5]:

tiles = bs.find_all(class_='goal-tile-wrapper')
tiles


# In[6]:

tile = tiles[0]
tile


# In[7]:

link = tile.findNext('a')
img = link.findNext('img')
img_title = img['alt'][:-5]
img_src = img['src']
link_href = link['href']
{'name': img_title, 'img_src': img_src, 'href': link_href}


# In[8]:

import collections
def get_data_from_goal_tile_wrapper_div(node, n=None):
    link = node.findNext('a')
    img = link.findNext('img')
    img_title = img['alt'][:-5]
    img_src = img['src']
    link_href = link['href']
    output = collections.OrderedDict({'@type': 'GlobalGoal'})
    if n:
        output['n'] = n
    output['name'] = img_title
    output['image'] = img_src
    output['url'] = link_href
    return output

def get_goal_tile_data(bs):
    for i, tile in enumerate(bs.find_all(class_='goal-tile-wrapper'), 1):
        yield get_data_from_goal_tile_wrapper_div(tile, n=i) 
        
tiles = list(get_goal_tile_data(bs))
import json
print(json.dumps(tiles, indent=2))


# In[9]:

goal_tiles = tiles[:-1]
goal_tiles


# In[ ]:





globalgoals-pyglobalgoals.py.py

# coding: utf-8

# # globalgoals-pyglobalgoals.py.ipynb
# 
# A jupyter notebook demonstrating HTML parsing, JSON-LD, RDFa, and schema.org in order to add Schema.org RDFa markup to @TheGlobalGoals for Sustainable Development at http://www.globalgoals.org/

# ## Background
# 
# * Homepage: **http://www.globalgoals.org/**
# - Twitter: https://twitter.com/TheGlobalGoals
# - Twitter: https://twitter.com/GlobalGoalsUN
# - Instagram: https://instagram.com/TheGlobalGoals/
# - Facebook: https://www.facebook.com/globalgoals.org
# - YouTube: https://www.youtube.com/channel/UCRfuAYy7MesZmgOi1Ezy0ng/
# - Hashtag: **#GlobalGoals**
#   - https://twitter.com/hashtag/GlobalGoals
#   - https://instagram.com/explore/tags/GlobalGoals/
#   - https://www.facebook.com/hashtag/GlobalGoals
# - Hashtag: #TheGlobalGoals
#   - https://twitter.com/hashtag/TheGlobalGoals
#   - https://instagram.com/explore/tags/TheGlobalGoals/
#   - https://www.facebook.com/hashtag/TheGlobalGoals
# - Hashtag: #Goal17
#   - https://twitter.com/hashtag/Goal17
#   - https://instagram.com/explore/tags/Goal17/
#   - https://www.facebook.com/hashtag/Goal17
# - Hashtag: #GlobalGoal17
#   - https://twitter.com/hashtag/GlobalGoal17
#   - https://instagram.com/explore/tags/GlobalGoal17/
#   - https://www.facebook.com/hashtag/GlobalGoal17
# 
# 
# ### pyglobalgoals
# 
# * Homepage: https://github.com/westurner/pyglobalgoals
# * Src: https://github.com/westurner/pyglobalgoals
# * Download: https://github.com/westurner/pyglobalgoals/releases
# 
# ### Objectives
# 
# * [x] ENH: Read and parse TheGlobalGoals from globalgoals.org
# * [x] ENH: Download (HTTP GET) each GlobalGoal tile image to ``./notebooks/data/images/``
# * [-] ENH: Generate e.g. tweets for each GlobalGoal (e.g. **##gg17** / **##GG17**)
# * [x] ENH: Save TheGlobalGoals to a JSON-LD document
# * [-] ENH: Save TheGlobalGoals with Schema.org RDF vocabulary (as JSON-LD)
# * [-] ENH: Save TheGlobalGoals as ReStructuredText with headings and images
# * [-] ENH: Save TheGlobalGoals as Markdown with headings and images
# * [-] ENH: Save TheGlobalGoals as RDFa with headings and images
# * [ ] ENH: Save TheGlobalGoals as RDFa with images like http://globalgoals.org/
# * [-] DOC: Add narrative documentation where necessary
# * [-] REF: Refactor and extract methods from ``./notebooks/`` to ``./pyglobalgoals/``
#   
# ## Implementation
# 
# * Python package: [**pyglobalgoals**](#pyglobalgoals)
# 
# * Jupyter notebook: **``./notebooks/globalgoals-pyglobalgoals.py.ipynb``**
#   * Src: https://github.com/westurner/pyglobalgoals/blob/master/notebooks/globalgoals-pyglobalgoals.py.ipynb
#   * Src: https://github.com/westurner/pyglobalgoals/blob/master/notebooks/globalgoals-pyglobalgoals.py.py
#   * Src: https://github.com/westurner/pyglobalgoals/blob/develop/notebooks/globalgoals-pyglobalgoals.py.ipynb
#   * Src: https://github.com/westurner/pyglobalgoals/blob/v0.1.2/notebooks/globalgoals-pyglobalgoals.py.ipynb
#   * Src: https://github.com/westurner/pyglobalgoals/blob/v0.2.1/notebooks/globalgoals-pyglobalgoals.py.ipynb
# 
#   * [x] Download HTML with requests
#   * [x] Parse HTML with beautifulsoup
#   * [x] Generate JSON[-LD] with ``collections.OrderedDict``
#   * [-] REF: Functional methods -> more formal type model -> ``pyglobalgoals.<...>``
# 
# 
# * [JSON-LD](#JSONLD) document: **``./notebooks/data/globalgoals.jsonld``**
#   * Src: https://github.com/westurner/pyglobalgoals/blob/master/notebooks/data/globalgoals.jsonld
# 
# 
# ### JSON-LD
# 
# * Wikipedia: https://en.wikipedia.org/wiki/JSON-LD
# * Homepage: http://json-ld.org/
# * Docs: http://json-ld.org/playground/
# * Hashtag: #JSONLD
# 
# ### RDFa
# 
# * Wikipedia: https://en.wikipedia.org/wiki/RDFa
# * Standard: http://www.w3.org/TR/rdfa-core/
# * Docs: http://www.w3.org/TR/rdfa-primer/
# * Hashtag: #RDFa

# In[1]:

#!conda install -y beautiful-soup docutils jinja2 requests
get_ipython().system(u"pip install -U beautifulsoup4 jinja2 'requests<2.8' requests-cache version-information # tweepy")


import bs4
import jinja2
import requests
import requests_cache

requests_cache.install_cache('pyglobalgoals_cache')

#!pip install -U version_information
get_ipython().magic(u'load_ext version_information')
get_ipython().magic(u'version_information jupyter, bs4, jinja2, requests, requests_cache, version_information')


# In[2]:

url = "http://www.globalgoals.org/"
req = requests.get(url)

#print(req)
#print(sorted(dir(req)))
#req.<TAB>
#req??<[Ctrl-]Enter>

if not req.ok:
    raise Exception(req)
content = req.content
print(content[:20])


# In[ ]:




# In[3]:

bs = bs4.BeautifulSoup(req.content)
print(bs.prettify())


# In[4]:

tiles = bs.find_all(class_='goal-tile-wrapper')
pp(tiles)


# In[5]:

tile = tiles[0]
print(tile)
print('##')
print(tile.prettify())


# In[6]:

link = tile.findNext('a')
img = link.findNext('img')
img_title = img['alt'][:-5]
img_src = img['src']
link_href = link['href']
example = {'name': img_title, 'img_src': img_src, 'href': link_href}
print(example)


# In[7]:

import collections
def get_data_from_goal_tile_wrapper_div(node, n=None):
    link = node.findNext('a')
    img = link.findNext('img')
    img_title = img['alt'][:-5]
    img_src = img['src']
    link_href = link['href']
    output = collections.OrderedDict({'@type': 'un:GlobalGoal'})
    if n:
        output['n'] = n
    output['name'] = img_title
    output['image'] = img_src
    output['url'] = link_href
    return output

def get_goal_tile_data(bs):
    for i, tile in enumerate(bs.find_all(class_='goal-tile-wrapper')[:-1], 1):
        try:
            yield get_data_from_goal_tile_wrapper_div(tile, n=i)
        except KeyError as e:
            print((i, tile))
            raise
        
tiles = list(get_goal_tile_data(bs))
import json
print(json.dumps(tiles, indent=2))
goal_tiles = tiles


# In[ ]:




# In[8]:


import codecs
from path import Path

def build_default_context():
    context = collections.OrderedDict()
    # context["dc"] = "http://purl.org/dc/elements/1.1/"
    context["schema"] = "http://schema.org/"
    # context["xsd"] = "http://www.w3.org/2001/XMLSchema#"
    # context["ex"] = "http://example.org/vocab#"
    # context["ex:contains"] = {
    #    "@type": "@id"
    # }
    
    # default attrs (alternative: prefix each with schema:)
    #  schema.org/Thing == schema:Thing (!= schema:thing)
    context["name"] = "http://schema.org/name"
    context["image"] = {
        "@type": "@id",
        "@id": "http://schema.org/image"
    }
    context["url"] = {
        "@type": "@id",
        "@id":"http://schema.org/url"
    }
    context["description"] = {
        "@type": "http://schema.org/Text",
        "@id": "http://schema.org/description"
    }
    return context

DEFAULT_CONTEXT = build_default_context()

def goal_tiles_to_jsonld(nodes, context=None, default_context=DEFAULT_CONTEXT):
    data = collections.OrderedDict()
    if context is None and default_context is not None:
        data['@context'] = build_default_context()
    elif context:
        data['@context'] = context
    elif default_context:
        data['@context'] = default_context
    data['@graph'] = nodes
    
    return data



DATA_DIR = Path('.') / 'data'
#DATA_DIR = Path(__file__).dirname
#DATA_DIR = determine_path_to(current_notebook) # PWD initially defaults to nb.CWD
DATA_DIR.makedirs_p()
GLOBAL_GOALS_JSONLD_PATH = DATA_DIR / 'globalgoals.jsonld'

def write_global_goals_jsonld(goal_tiles, path=GLOBAL_GOALS_JSONLD_PATH):
    goal_tiles_jsonld = goal_tiles_to_jsonld(goal_tiles)
    with codecs.open(path, 'w', 'utf8') as fileobj:
        json.dump(goal_tiles_jsonld, fileobj, indent=2)

def read_global_goals_jsonld(path=GLOBAL_GOALS_JSONLD_PATH, prettyprint=True):
    with codecs.open(path, 'r', 'utf8') as fileobj:
        global_goals_dict = json.load(fileobj,
                                     object_pairs_hook=collections.OrderedDict)
    return global_goals_dict

def print_json_dumps(global_goals_dict, indent=2):
    print(json.dumps(global_goals_dict, indent=indent))

    
write_global_goals_jsonld(goal_tiles)

global_goals_dict = read_global_goals_jsonld(path=GLOBAL_GOALS_JSONLD_PATH)

assert global_goals_dict == goal_tiles_to_jsonld(goal_tiles)

print_json_dumps(global_goals_dict)


# In[9]:

def build_tweet_for_goal_tile(node):
     return '##gg{n} {name} {url} {image} @TheGlobalGoals #GlobalGoals'.format(**node)
        
tweets = list(build_tweet_for_goal_tile(tile) for tile in goal_tiles)
tweets


# In[10]:


for node in goal_tiles:
    img_basename = node['image'].split('/')[-1]
    node['image_basename'] = img_basename
    node['tweet_txt'] = build_tweet_for_goal_tile(node)
    
print(json.dumps(goal_tiles, indent=2))


# In[11]:

#!conda install -y pycurl
try:
    import pycurl
except ImportError as e:
    import warnings
    warnings.warn(unicode(e))
def pycurl_download_file(url, dest_path, follow_redirects=True):
    with open(dest_path, 'wb') as f:
        c = pycurl.Curl()
        c.setopt(c.URL, url)
        c.setopt(c.WRITEDATA, f)
        if follow_redirects:
            c.setopt(c.FOLLOWLOCATION, True)
        c.perform()
        c.close()
    return (url, dest_path)


# In[12]:

import requests
def requests_download_file(url, dest_path, **kwargs):
    local_filename = url.split('/')[-1]
    # NOTE the stream=True parameter
    r = requests.get(url, stream=True)
    with open(dest_path, 'wb') as f:
        for chunk in r.iter_content(chunk_size=1024): 
            if chunk: # filter out keep-alive new chunks
                f.write(chunk)
                f.flush()
    return (url, dest_path)


# In[13]:

import urllib
def urllib_urlretrieve_download_file(url, dest_path):
    """
    * https://docs.python.org/2/library/urllib.html#urllib.urlretrieve
    """
    (filename, headers) = urlllib.urlretrieve(url, dest_path)
    return (url, filename)


# In[14]:


def deduplicate_on_attr(nodes, attr='image_basename'):
    attrindex = collections.OrderedDict()
    for node in nodes:
        attrindex.setdefault(node[attr], [])
        attrindex[node[attr]].append(node)
    return attrindex
        
def check_for_key_collisions(dict_of_lists):
    for name, _nodes in dict_of_lists.items():
        if len(_nodes) > 1:
            raise Exception(('duplicate filenames:')
                (name, nodes))
    
attrindex = deduplicate_on_attr(goal_tiles, attr='image_basename')
check_for_key_collisions(attrindex)
#

IMG_DIR = DATA_DIR / 'images'
IMG_DIR.makedirs_p()

def download_goal_tile_images(nodes, img_path):
    for node in nodes:
        dest_path = img_path / node['image_basename']
        source_url = node['image']
        (url, dest) = requests_download_file(source_url, dest_path)
        node['image_path'] = dest
        print((node['n'], node['name']))
        print((node['image_path']))
        # time.sleep(1)  # see: requests_cache
    
download_goal_tile_images(goal_tiles, IMG_DIR)

tiles_jsonld = goal_tiles_to_jsonld(goal_tiles)
print(json.dumps(tiles_jsonld, indent=2))


# In[15]:

#import jupyter.display as display
import IPython.display as display
display.Image(goal_tiles[0]['image_path'])


# In[16]:

import IPython.display
for tile in goal_tiles:
    x = IPython.display.Image(tile['image_path'])
x


# In[17]:

import IPython.display
def display_goal_images():
    for tile in goal_tiles:
        yield IPython.display.Image(tile['image_path'])
x = list(display_goal_images())
#pp(x)
IPython.display.display(*x)


# In[18]:


import string
print(string.punctuation)

NOT_URI_CHARS = dict.fromkeys(string.punctuation + string.digits)
NOT_URI_CHARS.pop('-')
NOT_URI_CHARS.pop('_')
def _slugify(txt):
    """an ~approximate slugify function for human-readable URI #fragments"""
    txt = txt.strip().lower()
    chars = (
        (c if c != ' ' else '-') for c in txt if
             c not in NOT_URI_CHARS)
    return u''.join(chars)

def _slugify_single_dash(txt):
    """
    * unlike docutils, this function does not strip stopwords like 'and' and 'or'
    TODO: locate this method in docutils
    """
    def _one_dash_only(txt):
        count = 0
        for char in txt:
            if char == '-':
                count += 1
            else:
                if count:
                    yield '-'
                yield char
                count = 0
    return u''.join(_one_dash_only(_slugify(txt)))
    
            

for node in goal_tiles:
    node['name_numbered'] = "%d. %s" % (node['n'], node['name'])
    node['slug_rst'] = _slugify_single_dash(node['name'])
    node['slug_md'] = _slugify_single_dash(node['name'])
    
print_json_dumps(goal_tiles)


# In[19]:

import IPython.display
def display_goal_images():
    for tile in goal_tiles:
        yield IPython.display.Markdown("## %s" % tile['name_numbered'])
        yield IPython.display.Image(tile['image_path'])
        yield IPython.display.Markdown(tile['tweet_txt'].replace('##', '\##'))
x = list(display_goal_images())
#pp(x)
IPython.display.display(*x)


# In[20]:

TMPL_RST = """

The Global Goals
******************

.. contents::

{% for node in nodes %}
{{ node['name_numbered'] }}
======================================================
| {{ node['url'] }}

.. image:: {{ node['image'] }}{# node['image_path'] #}
   :target: {{ node['url'] }}
   :alt: {{ node['name'] }} 

..

   {{ node['tweet_txt'] }}
   
{% endfor %}
"""
tmpl_rst = jinja2.Template(TMPL_RST)
output_rst = tmpl_rst.render(nodes=goal_tiles)
print(output_rst)



# In[21]:

output_rst_path = DATA_DIR / 'globalgoals.rst'
with codecs.open(output_rst_path, 'w', encoding='utf-8') as f:
    f.write(output_rst)
    print("# wrote goals to %r" % output_rst_path)


# In[22]:

import docutils.core
output_rst_html = docutils.core.publish_string(output_rst, writer_name='html')
print(bs4.BeautifulSoup(output_rst_html).find(id='the-global-goals'))


# In[23]:

IPython.display.HTML(output_rst_html)


# In[24]:


TMPL_MD = """

# The Global Goals

**Contents:**
{% for node in nodes %}
* [{{ node['name_numbered'] }}](#{{ node['slug_md'] }})
{%- endfor %}

{% for node in nodes %}
## {{ node['name_numbered'] }}
{{ node['url'] }}

[![{{node['name_numbered']}}]({{ node['image'] }})]({{ node['url'] }})

> {{ node['tweet_txt'] }}

{% endfor %}
"""
tmpl_md = jinja2.Template(TMPL_MD)
output_markdown = tmpl_md.render(nodes=goal_tiles)
print(output_markdown)


# In[25]:

output_md_path = DATA_DIR / 'globalgoals.md'
with codecs.open(output_md_path, 'w', encoding='utf-8') as f:
    f.write(output_markdown)
    print("# wrote goals to %r" % output_md_path)


# In[26]:

IPython.display.Markdown(output_markdown)


# In[27]:

context = dict(nodes=goal_tiles)


# In[28]:

TMPL_HTML = """

<h1>The Global Goals</h1>

<h2>Contents:</h2>
{% for node in nodes %}
<li><a href="#{{node.slug_md}}">{{node.name_numbered}}</a></li>
{%- endfor %}
{% for node in nodes %}
<div class="goal-tile">
  <h2><a name="#{{node.slug_md}}">{{ node.name_numbered }}</a></h2>
  <a href="{{node.url}}">{{node.url}} </a>
  <a href="{{node.url}}">
    <img src="{{node.image}}" alt="{{node.name_numbered}}"/>{{node.url}} </a>
  <div style="margin-left: 12px">
  {{ node.tweet_txt }}
  </div>
</div>
{% endfor %}
"""
tmpl_html = jinja2.Template(TMPL_HTML)
output_html = tmpl_html.render(**context)
print(output_html)


# In[29]:

output_html_path = DATA_DIR / 'globalgoals.html'
with codecs.open(output_html_path, 'w', encoding='utf-8') as f:
    f.write(output_html)
    print("# wrote goals to %r" % output_html_path)


# In[30]:

IPython.display.HTML(output_html)


# In[31]:

import jinja2
# TODO: prefix un:
TMPL_RDFA_HTML5 = ("""
<div prefix="schema: http://schema.org/
             un: http://schema.un.org/#">
<h1>The Global Goals</h1>
<h2>Contents:</h2>
{%- for node in nodes %}
<li><a href="#{{node.slug_md}}">{{node.name_numbered}}</a></li>
{%- endfor %}
{% for node in nodes %}
<div class="goal-tile" resource="{{node.url}}" typeof="un:GlobalGoal">
  <div style="display:none">
    <meta property="schema:name">{{node.name}}</meta>
    <meta property="schema:image">{{node.image}}</meta>
    <meta property="#n">{{node.n}}</meta>
  </div>
  <h2><a name="#{{node.slug_md}}">{{ node.name_numbered }}</a></h2>
  <a property="schema:url" href="{{node.url}}">{{node.url}} </a>
  <a href="{{node.url}}">
    <img src="{{node.image}}" alt="{{node.name_numbered}}"/>{{node.url}} </a>
  <div style="margin-left: 12px">
  {{ node.tweet_txt }}
  </div>
</div>
{% endfor %}
</div>
"""
)
tmpl_rdfa_html5 = jinja2.Template(TMPL_RDFA_HTML5)
output_rdfa_html5 = tmpl_rdfa_html5.render(**context)
print(output_rdfa_html5)


# In[32]:

output_rdfa_html5_path = DATA_DIR / 'globalgoals.rdfa.html5.html'
with codecs.open(output_rdfa_html5_path, 'w', encoding='utf-8') as f:
    f.write(output_rdfa_html5_path)
    print("# wrote goals to %r" % output_rdfa_html5_path)


# In[33]:

IPython.display.HTML(output_rdfa_html5)


# In[34]:

# tmpl_html
# tmpl_rdfa_html5
import difflib
for line in difflib.unified_diff(
    TMPL_HTML.splitlines(),
    TMPL_RDFA_HTML5.splitlines()):
    print(line)

